File: day2/part1.jl
module Part1

using Base: RegexMatchIterator
export compute

function parse_file(filename::String)::Vector{String}
    file = open(filename)
    lines = Vector{String}()
    for line in readlines(file)
        push!(lines, line)
    end
    close(file)
    return lines
end

function extract_gameinfo(filename::String)::Vector{Vector{String}}
    lines = parse_file(filename)
    out = Vector{Vector{String}}()
    pattern = Regex("[0-9]+|red|green|blue")
    for line in lines
        result = [match.match for match in eachmatch(pattern, line)]
        push!(out, result)
    end
    return out
end

function is_possible(game::Vector{String}, limits::Dict{String, Int})::Int
    id = parse(Int, game[1])
    picks = game[2:end]
    iter_length::Int = length(picks)/2
    for i in 1:iter_length
        number = parse(Int, picks[2*i-1])
        value = picks[2*i]
        if number > limits[value]
            return 0
        end
    end
    return id
end

function compute(filename::String, limits::Dict{String, Int})::Int
    games = extract_gameinfo(filename)
    sum = 0
    for game in games
        sum += is_possible(game, limits)
    end
    return sum
end

end

File: day2/part1.py
import re
import sys

class Part1:
    def __init__(self, filename: str, limits: dict[str, int]) -> None:
        self.filename = filename    
        self.limits = limits

    def parse_file(self) -> list[str]:
        with open(self.filename, 'r') as file:
            out = file.readlines()
        return out

    def extract_gameinfo(self) -> list[list[str]]:
        lines = self.parse_file()
        out = []
        for line in lines:
            pattern = re.compile(r"[0-9]+|red|green|blue")
            result = pattern.findall(line)
            out.append(result)
        return out

    def is_possible(self, game: list[str]) -> int:
        id = game[0]
        picks = game[1::]
        iter_len = int(len(picks)/2)
        for i in range(iter_len):
            number = int(picks[2*i])
            value = picks[2*i+1]
            if number > self.limits[value]:
                return 0
        return int(id)
    
    def compute(self) -> int:
        games = self.extract_gameinfo()
        sum = 0
        for game in games:
            sum += self.is_possible(game)
        return sum


def main() -> None:
    try:
        filename = sys.argv[1]
    except IndexError:
        print(f"No filename input. using default \"input.txt\"")
        filename = "input.txt"
    limits = {"red" : 12, "green" : 13, "blue" : 14}
    soln = Part1(filename, limits)
        
    answer = soln.compute()
    print(answer)


if __name__ == "__main__":
    main()


